pkg/reconciler/nscert/nscert.go

```diff
@@ -0,0 +1,204 @@
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+	ctx = c.configStore.ToContext(ctx)
+
+	_, ns, err := cache.SplitMetaNamespaceKey(key)
+	original, err := c.nsLister.Get(ns)
+
+	if apierrs.IsNotFound(err) {
+		logger.Errorf("Namespace %s in work queue no longer exists %s", key, err)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	newNamespace := original.DeepCopy()
+
+	_, ok := newNamespace.Labels[networking.WildcardCertLabelKey]
+	if ok {
+		logger.Infof("Skipping wildcard certificate creation for excluded namespace %s", newNamespace.Name)
+		return nil
+	}
+
+	err = c.reconcile(ctx, newNamespace)
+	if err != nil {
+		c.Recorder.Event(newNamespace, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	cfg := config.FromContext(ctx)
+
+	// Only create wildcard certs for the default domain
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
+	if defaultDomain == "" {
+		c.Logger.Infof("Skipping wildcard certificate creation for namespace %s. No default domain defined.", ns.Name)
+		return nil
+	}
+
+	dnsName, err := wildcardDomain(cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	if err != nil {
+		c.Logger.Errorf("Failed to apply domain template %s to domain %s and namespace %s: %v", cfg.Network.DomainTemplate, defaultDomain, ns.Name, err)
+		return err
+	}
+
+	desiredCert := resources.MakeWildcardCertificate(ns, dnsName, defaultDomain)
+	labelSelector := kubelabels.SelectorFromSet(
+		kubelabels.Set{
+			networking.WildcardCertDomainLabelKey: defaultDomain,
+		},
+	)
+
+	// If any labeled cert has been issued for our DNSName then there's nothing to do
+	existingCerts, err := c.knCertificateLister.Certificates(ns.Name).List(labelSelector)
+	if err != nil {
+		return fmt.Errorf("Failed to list certificates: %v", err)
+	}
+
+	matchingCert := findMatchingCert(dnsName, existingCerts)
+	if matchingCert != nil {
+		return nil
+	}
+
+	// If there is no matching cert find one previously created by this reconciler which may
+	// need to be updated.
+	existingCert, err := findNamespaceCert(ns, existingCerts)
+
+	if apierrs.IsNotFound(err) {
+		cert, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).Create(desiredCert)
+		if err != nil {
+			c.Logger.Errorw("Failed to create namespace certificate", zap.Error(err))
+			c.Recorder.Eventf(ns, corev1.EventTypeWarning, "CreationFailed",
+				"Failed to create Knative certificate %s/%s: %v", ns.Name, desiredCert.ObjectMeta.Name, err)
+			return err
+		}
+
+		c.Recorder.Eventf(cert, corev1.EventTypeNormal, "Created",
+			"Created Knative Certificate %s/%s", ns.Name, cert.ObjectMeta.Name)
+	} else if err != nil {
+		return err
+	} else if !metav1.IsControlledBy(existingCert, ns) {
+		existingCert.Status.MarkResourceNotOwned("CertManagerCertificate", existingCert.Name)
```

Comment:
Generally this would be reflected in the parent resource's status.  It is weird that we would mutate another resources status (though it seems to only in-memory and not persist back to the API?).  Generally we don't mutate the status of resources other than the primary key of a given reconciler.

Reply:
Even though this reconciler creates knative certs it shouldn't modify their statuses? In the case where the wildcard cert's owner has changed should this reconciler just reset the ownerref?